File: src/main/java/tk/hiddenname/smarthome/model/task/trigger/objects/TriggerObject.java
package tk.hiddenname.smarthome.model.task.trigger.objects;

import com.fasterxml.jackson.annotation.JsonTypeInfo;
import com.fasterxml.jackson.databind.annotation.JsonTypeIdResolver;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import tk.hiddenname.smarthome.model.task.trigger.TriggerAction;

import javax.persistence.*;
import javax.validation.constraints.NotNull;

@Data
@Entity
@NoArgsConstructor
@EqualsAndHashCode(of = {"id"})
@Table(name = "trigger_object")
@Inheritance(strategy = InheritanceType.JOINED)
@JsonTypeInfo(use = JsonTypeInfo.Id.CUSTOM, include = JsonTypeInfo.As.EXISTING_PROPERTY, property = "action", visible = true)
@JsonTypeIdResolver(TriggerObjectTypeIdResolver.class)
public abstract class TriggerObject {

    @Id
    @GeneratedValue(strategy = GenerationType.SEQUENCE)
    @Column(updatable = false, nullable = false)
    private Integer id;

    @NotNull
    @Column(nullable = false, updatable = false)
    @Enumerated(EnumType.STRING)
    private TriggerAction action;
}

File: src/main/java/tk/hiddenname/smarthome/model/task/trigger/objects/TriggerObject.kt
package tk.hiddenname.smarthome.model.task.trigger.objects

import com.fasterxml.jackson.annotation.JsonTypeInfo
import com.fasterxml.jackson.databind.annotation.JsonTypeIdResolver
import tk.hiddenname.smarthome.model.AbstractJpaPersistable
import tk.hiddenname.smarthome.model.task.trigger.TriggerAction
import javax.persistence.*

@Entity
@Table(name = "trigger_object")
@Inheritance(strategy = InheritanceType.JOINED)
@JsonTypeInfo(use = JsonTypeInfo.Id.CUSTOM, include = JsonTypeInfo.As.EXISTING_PROPERTY, property = "action", visible = true)
@JsonTypeIdResolver(TriggerObjectTypeIdResolver::class)
abstract class TriggerObject(
        @Column(updatable = false, nullable = false)
        @Enumerated(EnumType.STRING)
        open val action: TriggerAction = TriggerAction.NOT_SPECIFIED
) : AbstractJpaPersistable()